### PR TITLE
OBJFileLoader: Fix AssetContainer handling and add opt-in texture loading waits

### PR DESCRIPTION
**Pull request by noname0310**
## Summary

  This PR fixes resource collection and ownership in the OBJ loader and adds an option to wait for referenced textures before completing a load.

  ## Changes

  - Return created geometries from `importMeshAsync` and include them in `AssetContainer.geometries`.
  - Prevent container-owned geometries, textures, and generated point/line materials from being registered in the scene prematurely, and assign their `_parentContainer`.
  - Skip unused MTL materials and their textures instead of creating and subsequently disposing unused materials.
  - Reuse existing textures when cloning materials for line rendering.
  - Preserve line markers until all line materials have been processed to avoid unnecessary material cloning.
  - Honor the per-load `invertTextureY` option.

  ## Optional texture loading waits

  Adds `OBJLoadingOptions.waitForTextures` defaulting to `false` to preserve the existing non-waiting behavior.

  When enabled:
  - Referenced texture loading promises are awaited.
  - Delayed textures are started without requiring the container to be rendered.
  - Texture loading failures follow `materialLoadingFailsSilently`.

  ```ts
  const container = await LoadAssetContainerAsync("model.obj", scene, {
      pluginOptions: {
          obj: {
              waitForTextures: true,
              materialLoadingFailsSilently: false,
          },
      },
  });
  ```

 ## Validation

  - 29 OBJ unit tests passed, covering resource ownership, shared textures, optional waiting, delayed loading, and error handling.

Note: The tests were AI-generated, but I have not reviewed their code quality.

### CODE REVIEW

File: packages/dev/loaders/src/OBJ/solidParser.ts [hunk 1]
```diff
@@ -921,7 +921,10 @@ export class SolidParser {
 
                 if (!this._materialNameFromObj) {
                     // Create a material with point cloud on
+                    scene._blockEntityCollection = !!assetContainer;
                     newMaterial = new StandardMaterial(Geometry.RandomId(), scene);
+                    newMaterial._parentContainer = assetContainer;
+                    scene._blockEntityCollection = false;
```
[the remark this answers is not in the document]
Comment (noname0310) [reply packages/dev/loaders/src/OBJ/solidParser.ts:927]: I'm tired of seeing these pointless AI reviews...

**Review by RaananW — CHANGES_REQUESTED** [review]
> 🤖 This review comment was created by an AI agent.

Requesting changes for one new correctness blocker:

- [`OBJFileLoader._parseSolidAsync` (`objFileLoader.pure.ts:356-363`)](https://github.com/BabylonJS/Babylon.js/blob/b483c23f95b15ce19d70836aae28d5ea8576a139/packages/dev/loaders/src/OBJ/objFileLoader.pure.ts#L356-L363): with `waitForTextures: true` and the default `materialLoadingFailsSilently: true`, `Promise.all` rejects as soon as any texture fails; the catch immediately resolves the outer MTL promise, so the OBJ load can complete while other referenced textures are still pending. This violates the option’s wait contract. In silent mode, catch each texture rejection (or use `Promise.allSettled`) and still await every texture; add a multi-texture regression test where one fails before another settles.

Existing review threads: Copilot’s `solidParser.ts` restoration concern remains valid and unresolved, so I did not duplicate it inline. Its `mtlFileLoader.ts` comment describes behavior already present on the base commit and is not counted as PR-introduced.

CI: repository build/test/lint/typecheck checks have not run; only GitGuardian passed. This review requests changes for the code issue above, not unfinished CI.

**Comment by noname0310** [comment]
@RaananW 

The `_blockEntityCollection` restoration case mentioned here is also not handled by `GLTFFileLoader`. Its `_createDefaultMaterial` method follows the same pattern

see: https://github.com/BabylonJS/Babylon.js/blob/master/packages/dev/loaders/src/glTF/2.0/glTFLoader.pure.ts#L2415-L2432

OBJ does the same as glTF, so I’d leave this for a separate PR.

The waitForTextures issue has been fixed, with regression tests

File: packages/dev/loaders/src/OBJ/solidParser.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (RaananW) [reply packages/dev/loaders/src/OBJ/solidParser.ts:927]: this is actually valid. In [`SolidParser` lines 924–927](https://github.com/BabylonJS/Babylon.js/blob/6c47b07ce54521fa5d9be616827bf2857ddd1f45/packages/dev/loaders/src/OBJ/solidParser.ts#L924-L927), `_blockEntityCollection` is unconditionally reset to `false`. This is only safe if its previous value was false and material construction cannot throw - neither is guaranteed. The similar glTF implementation means glTF has the same pre-existing weakness; it does not justify introducing it in new OBJ code. The new line-material path at [`objFileLoader.pure.ts` lines 397–401](https://github.com/BabylonJS/Babylon.js/blob/6c47b07ce54521fa5d9be616827bf2857ddd1f45/packages/dev/loaders/src/OBJ/objFileLoader.pure.ts#L397-L401) also restores on success but lacks `try/finally`.

**Comment by RaananW** [comment]
Commented about the copilot comment. The older Copilot comment on `mtlFileLoader.ts` describes pre-existing code and should not block this PR

**Review by RaananW — CHANGES_REQUESTED** [review]
> 🤖 This review comment was created by an AI agent.

Re-review verdict: 🔴 Request changes

The follow-up commit fixes the previously reported `waitForTextures` contract: `Promise.allSettled` now waits for every referenced texture before propagating or silently handling failures, and the new two-texture tests cover both completion and failure of the remaining request.

One previously raised correctness blocker remains on the current head: `_blockEntityCollection` is still not restored with `try/finally` around the newly container-aware material constructors in [`SolidParser.parse` (`solidParser.ts:924-927`)](https://github.com/BabylonJS/Babylon.js/blob/6c47b07ce54521fa5d9be616827bf2857ddd1f45/packages/dev/loaders/src/OBJ/solidParser.ts#L924-L927) and [`OBJFileLoader._parseSolidAsync` (`objFileLoader.pure.ts:397-401`)](https://github.com/BabylonJS/Babylon.js/blob/6c47b07ce54521fa5d9be616827bf2857ddd1f45/packages/dev/loaders/src/OBJ/objFileLoader.pure.ts#L397-L401). A constructor failure leaves the scene-wide flag altered, while a pre-existing blocked state can be clobbered. The analogous MTL path predates this PR and is not counted as a new blocker. Please preserve the prior value and restore it in `finally` for these new paths.

CI: ⚠️ Missing coverage — only GitGuardian is reported; build, tests, lint/format, and type-check checks are absent.
CI safety: ✅ No credential-exposure risk found in the changed source/tests.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to the build configuration file.

*This build will release a new package on npm*

If that was unintentional please make sure to revert those changes or close this PR.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/18895/merge

Test environment:
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18895/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18895/merge/index.html#WGZLGJ#4600

Links to test your changes to core in the published versions of the Babylon tools (does not contain changes you made to the tools themselves):

https://playground.babylonjs.com/?snapshot=refs/pull/18895/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/18895/merge
https://gui.babylonjs.com/?snapshot=refs/pull/18895/merge
https://nme.babylonjs.com/?snapshot=refs/pull/18895/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/18895/merge#BCU1XR#0

*If you made changes to the sandbox or playground in this PR, additional comments will be generated soon containing links to the dev versions of those tools.*

**Comment by bjsplat** [comment]
Interaction tests

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18895/merge/testResults/interactionplaywright/index.html

**Comment by bjsplat** [comment]
## 🟢 Memory Leak Test Results

**4** passed, **0** leaked out of **4** scenarios

🟢 All memory leak tests passed — no leaks detected.

<details>
<summary>Passed Scenarios (4)</summary>

| Scenario | Package |
|----------|---------|
| Core Playground #2FDQT5#1508 | @babylonjs/core |
| Core Playground #T90MQ4#14 | @babylonjs/core |
| Core Playground #8EDB5N#2 | @babylonjs/core |
| Core Playground #LL5BIQ#636 | @babylonjs/core |

</details>

**Comment by bjsplat** [comment]
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18895/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
## ⚡ Performance Test Results

🟢 All performance tests passed — no regressions detected.

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18895/merge/testResults/webgl2playwright/index.html

**Comment by noname0310** [comment]
OBJ's `_blockEntityCollection` handling is intended to follow the existing glTF loader convention:
set it to `!!assetContainer` while creating or applying a resource, assign `_parentContainer`, and then reset it to `false`.

for instance: https://github.com/BabylonJS/Babylon.js/blob/deab6a2c18947c4c2988bafa1eddbdce46bd1851/packages/dev/loaders/src/glTF/2.0/glTFLoader.pure.ts#L2415-L2419

If anything, the code elsewhere that saves and restores the previous state appears to be the result of a loss of context.

**Review by deltakosh — CHANGES_REQUESTED** [review]
Hi! Thanks for the detailed loader work. I found two correctness issues in the new texture-waiting path:\n\n1. A second waitForTextures load can hang forever when it reuses an already-failed cached texture. The new wrapper subscribes after the cached texture's error notification has fired, so its deferred promise never settles. Please retry or replay the cached failure state and cover a second load of the same failed URL.\n\n2. With waitForTextures: true and materialLoadingFailsSilently: false, a texture failure rejects before the created meshes, geometries, materials, and textures are copied into the AssetContainer. Because scene collection was blocked and the catch only clears _assetContainer, those resources are orphaned and cannot be disposed by the caller. Please track or clean up all resources on this rejection path and add a reference-count regression test.\n\nThe earlier _blockEntityCollection restoration concern is not part of this review. The changelog label and full required CI are also still missing. Thanks for taking another pass at these edge cases!

**Review by RaananW — CHANGES_REQUESTED** [review]
> 🤖 This review comment was created by an AI agent.

Re-review verdict: 🔴 Request changes

The prior `_blockEntityCollection` blocker is still open on this head. The latest commit removes the previous-value snapshots and resets the scene-wide flag to `false`. That can still clobber a caller's existing blocked state, and the new point-cloud and line-material constructor paths still do not restore the flag if construction throws:

- [`SolidParser.parse` (`solidParser.ts:924-927`)](https://github.com/BabylonJS/Babylon.js/blob/6493880b9b5451135ab1c75606cfc2e0fa7049ab/packages/dev/loaders/src/OBJ/solidParser.ts#L924-L927)
- [`OBJFileLoader._parseSolidAsync` (`objFileLoader.pure.ts:397-400`)](https://github.com/BabylonJS/Babylon.js/blob/6493880b9b5451135ab1c75606cfc2e0fa7049ab/packages/dev/loaders/src/OBJ/objFileLoader.pure.ts#L397-L400)

Please preserve the previous value and restore it in `finally` for the new paths. The two texture-waiting issues raised in the current team review are also blocking and are not repeated here.

CI: ⚠️ Missing coverage — only GitGuardian is reported on this head; build, tests, lint/format, and type-check checks are absent.
CI safety: ✅ No credential-exposure risk found in the changed source and tests.

**Comment by noname0310** [comment]
@deltakosh I've fixed the cleanup when loading throws.

The missing onError callback when reusing a cached InternalTexture was identified as a bug and addressed in #11217 in 2021, but it appears to still persist.

Would it be okay if I opened a separate PR to fix it?

**Comment by bjsplat** [comment]
Interaction tests

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18895/merge/testResults/interactionplaywright/index.html

**Comment by bjsplat** [comment]
## 🟢 Memory Leak Test Results

**4** passed, **0** leaked out of **4** scenarios

🟢 All memory leak tests passed — no leaks detected.

<details>
<summary>Passed Scenarios (4)</summary>

| Scenario | Package |
|----------|---------|
| Core Playground #2FDQT5#1508 | @babylonjs/core |
| Core Playground #T90MQ4#14 | @babylonjs/core |
| Core Playground #8EDB5N#2 | @babylonjs/core |
| Core Playground #LL5BIQ#636 | @babylonjs/core |

</details>

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18895/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
## ⚡ Performance Test Results

🟢 All performance tests passed — no regressions detected.

**Review by RaananW — CHANGES_REQUESTED** [review]
🤖 This review comment was created by an AI agent.

Thanks for the cleanup update. Resources are now tracked as they are created, and a failed container is disposed. This fixes the earlier orphaned-resource concern.

I still need these changes:

1. The earlier `_blockEntityCollection` blocker is still open in the new point and line material paths. They replace the caller's prior value with `false`. A constructor error can also leave the value changed. Please save the old value and restore it in `finally`: [solidParser.ts:924-928](https://github.com/BabylonJS/Babylon.js/blob/efdd76b2a84205ad2f6e686841123e8685ae7a18/packages/dev/loaders/src/OBJ/solidParser.ts#L924-L928) and [objFileLoader.pure.ts:373-398](https://github.com/BabylonJS/Babylon.js/blob/efdd76b2a84205ad2f6e686841123e8685ae7a18/packages/dev/loaders/src/OBJ/objFileLoader.pure.ts#L373-L398).

2. `waitForTextures: true` can still wait forever on a second load of an already failed cached texture. `_GetTexture` only receives future callbacks, after the cached error was already sent. Please retry the failed texture or replay its error, and add a second-load regression test: [mtlFileLoader.ts:267-276](https://github.com/BabylonJS/Babylon.js/blob/efdd76b2a84205ad2f6e686841123e8685ae7a18/packages/dev/loaders/src/OBJ/mtlFileLoader.ts#L267-L276).

3. Required CI is not green. `Monorepo CI (Memory leak tests (packages suite))` and `Monorepo CI (Visualization tests - WebGL 2)` are failing. Performance and WebGPU checks are still pending.

CI safety: I found no credential exposure risk in the changed source and tests.

**Comment by bjsplat** [comment]
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18895/merge/testResults/webgpuplaywright/index.html

**Comment by RaananW** [comment]
any update here?

**Comment by noname0310** [comment]
it's ready to merge as is context. i think

File: packages/dev/loaders/src/OBJ/mtlFileLoader.ts [hunk 2]
```diff
@@ -69,6 +84,10 @@ export class MTLFileLoader {
                     //Add the previous material in the material array.
                     this.materials.push(material);
                 }
+                if (materialNames && !materialNames.has(value)) {
+                    material = null;
+                    continue;
+                }
```
[the remark this answers is not in the document]
Comment (RaananW) [reply packages/dev/loaders/src/OBJ/mtlFileLoader.ts:90]: [Responded by Copilot on behalf of RaananW] Fixed in f08f11511b. Every newly introduced OBJ loader path that temporarily changes `scene._blockEntityCollection` now captures the prior value and restores it in `finally`, including material, texture, mesh, geometry, generated line/point material, and cloned line material creation. Regression coverage verifies that a pre-existing blocked state survives both successful loading and clone failures.

File: packages/dev/loaders/src/OBJ/solidParser.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (RaananW) [reply packages/dev/loaders/src/OBJ/solidParser.ts]: [Responded by Copilot on behalf of RaananW] Fixed in f08f11511b. The state restoration now covers this mesh/geometry path as well as all related material and texture creation paths, including the additional line-material path called out in the discussion. Each path preserves the caller's original value through `try`/`finally`, with success and failure regression tests.

**Review by RaananW — APPROVED** [review]
The valid collection-state restoration issue is addressed in f08f11511b with focused success and failure regression coverage. The remaining prior CI failures were confirmed infrastructure failures (GitHub checkout authentication and exhausted BrowserStack capacity), so the pushed commit should provide a clean rerun.

**Review by RaananW — APPROVED** [review]
Approved on 14299c6be6. The collection-state blocker is fixed in f08f11511b, the cached failed-texture retry blocker is fixed in 14299c6be6, and the prior failed-container resource cleanup remains covered. Focused OBJ tests pass 32/32, loaders compile, targeted lint/format pass, and full repository lint/tree-shaking checks pass.

**Comment by RaananW** [comment]
@deltakosh Both correctness issues from your change request are now addressed: failed container resources are disposed with reference-count coverage, and 14299c6be6 evicts failed cached internal textures so a subsequent waitForTextures load retries instead of hanging. The _blockEntityCollection paths are also restored safely in f08f11511b. Could you please re-review the latest head once CI completes?

**Review by deltakosh — CHANGES_REQUESTED** [review]
Follow-up to the existing cached-failure blocker: eviction still depends on the first load using `waitForTextures: true`. With `EngineStore.UseFallbackTexture = false`, a non-waiting OBJ load that subsequently encounters a texture failure leaves the failed internal cached under its original URL. If that texture/container remains undisposed, a later waiting load using the same engine, URL, and texture options reuses the failed internal and subscribes after its error notification, so its deferred never settles.

Please evict failed entries independently of waiting mode, or detect/replay cached failures, and add a non-waiting-failure → waiting-load regression. Thanks for continuing to tighten this path!

**Review by RaananW — APPROVED** [review]
Approved on 475b270841. Failed internal textures are now evicted regardless of whether the originating load waits, and the regression covers both non-waiting and waiting first loads followed by a waiting retry. The two CI test-lint violations are also fixed. OBJ tests pass 33/33, loaders compile, explicit source/test ESLint and formatting pass, and repository lint/cycles/tree-shaking checks pass.

**Comment by RaananW** [comment]
@deltakosh Addressed your latest review in 475b270841: the texture error callback now evicts the failed internal cache entry independently of waitForTextures, and the regression is parameterized to cover both a non-waiting failure and a waiting failure followed by a waiting retry. I also fixed the two vitest/valid-expect errors from CI; the Formatting task itself had passed. Could you please re-review the latest head?

**Review by deltakosh — APPROVED** [review]
The failed-texture eviction now applies to both waiting and non-waiting originating loads, and the parameterized retry regression covers the previously hanging sequence. Exact CI is fully green. Approved—thanks for the thorough follow-up!

**Comment by bjsplat** [comment]
Interaction tests

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18895/merge/testResults/interactionplaywright/index.html

**Comment by bjsplat** [comment]
## 🟢 Memory Leak Test Results

**4** passed, **0** leaked out of **4** scenarios

🟢 All memory leak tests passed — no leaks detected.

<details>
<summary>Passed Scenarios (4)</summary>

| Scenario | Package |
|----------|---------|
| Core Playground #2FDQT5#1508 | @babylonjs/core |
| Core Playground #T90MQ4#14 | @babylonjs/core |
| Core Playground #8EDB5N#2 | @babylonjs/core |
| Core Playground #LL5BIQ#636 | @babylonjs/core |

</details>

**Comment by bjsplat** [comment]
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18895/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18895/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
## ⚡ Performance Test Results

🟢 All performance tests passed — no regressions detected.

**Comment by noname0310** [comment]
@RaananW @deltakosh I want to reiterate two points I already raised, because the follow-up changes do not address the reasoning behind my earlier comments.

### `_blockEntityCollection`

Resetting this flag to `false` was intentional, not an oversight. As I explained earlier, this follows the existing glTF loader convention: set the flag for the resource operation, assign `_parentContainer`, and reset it afterward.

I consider the insistence on preserving the previous value a misinterpretation of that context. This looks like an agent applying a generic "save and restore mutable state" pattern without accounting for how the flag is actually used by the loaders. 

If this convention needs to change, it should be discussed and handled consistently across loaders in a separate PR, rather than classified as an OBJ-specific correctness defect.

### Cached texture error callbacks

The missing `onError` callback when reusing an already-failed cached `InternalTexture` is a bug in the texture loader. It should be fixed there, not worked around in the OBJ loader.

The fact that `waitForTextures` exposes the problem does not make the OBJ loader the appropriate place to fix it. Error propagation for cached textures should behave correctly for all consumers, without each loader implementing its own cache-recovery logic.


I would appreciate a maintainer addressing these points directly. Repeated agent-generated requests that do not engage with the explanations already provided are not resolving the disagreement.

**Comment by RaananW** [comment]
Hey,

Will address your concerns. Writing this message myself, i promise :)

First - the fact that the glTF loader does something does not mean it is the right behavior. It might mean that it needs to be fixed there. This can be a future PR (that I am already working on). I know that Copilot comments can sometimes feel redundant, and can feel out of place. And I tend to take them with a grain of salt. But this is actually a valid concern. Yo uare right - this is a change in the way we deal with the flag.

About the onError - another valid concern. However, i agree with you - this was a small "fight" between - do we need it, and where does it belong. And this is a perfect followup - move that to a higher lvel and allow every user of the feature to enjoy the benefits. It does not mean it shouldn't have been implemented. 

So, even though AI agents can be slightly rigid, they had some good points. They also allow us to accelerate our work. I know it sometimes feel "heart-less", but it's quite effective. I won't get into the philosophy of whether or not it is coding, or whether or not we should or should not use it. The fact is, that 99.9% of the PRs we are getting (including yours...) are AI generated. And we need to stay ahead.

I am working now to fix the follow ups.

Thanks a lot for the contribution. I really hope to see more PRs from you.
